In [3]:
import random
import time
import sys
sys.setrecursionlimit(8000)

In [4]:


class NTHUList(list):
    def selectSort(self): #O(N^2)
        for i in range(len(self)-1): #界線要搞清楚
            for j in range(i+1,len(self)):
                if self[i] < self[j]:
                    self[i],self[j] = self[j],self[i] #python 特殊語法
        return

    def bubbleSort(self): #O(N^2)
        for i in range(1,len(self)): #開始不可為0/長度不可為1
            j = i
            while j > 0 and self[j] > self[j-1]:
                self[j],self[j-1] = self[j-1],self[j]
                j = j-1
        return

    def mergeSort(self): 
        if len(self) <= 1:
            return self
        left = NTHUList(self[:len(self)//2]) #// is integer divide
        right = NTHUList(self[len(self)//2:]) #不用+1

        sortedLeft = left.mergeSort()
        sortedRight = right.mergeSort()

        returnList = []
        ri = 0
        li = 0
        while ri < len(sortedRight) and li < len(sortedLeft):
            if sortedLeft[li] > sortedRight[ri]:
                returnList.append(sortedLeft[li])
                li+=1
            else:
                returnList.append(sortedRight[ri])
                ri+=1
        if ri == len(sortedRight):
            returnList.extend(sortedLeft[li:]) #append 是一個 element extend 是一個 list
        else:
            returnList.extend(sortedRight[ri:])
        return returnList

    def heapSort(self):
        self.heapify() 
        end = len(self)-1
        while end > 0 :
            self[0],self[end] = self[end],self[0]
            end = end-1
            self.shiftdown(0,end)
        return

    def heapify(self):
        for i in range(len(self)):
            self.shiftup(i)
        return
    def shiftup(self,i):
        if (i == 0):
            return
        else:
            if (self[i] <= self[(i-1)//2]):
                return
            else :
                self[i],self[(i-1)//2] = self[(i-1)//2],self[i]
                self.shiftup((i-1)//2)
    def shiftdown(self,i,end):
        leftChild = 2*i+1
        rightChild = 2*i+2
        if (leftChild > end):
            return
        elif (leftChild == end):
            if (self[leftChild] > self[i]):
                self[i],self[leftChild] = self[leftChild],self[i]
            return
        else:
            if (self[i] < max(self[leftChild],self[rightChild])):
                if self[leftChild] > self[rightChild]:
                    self[i],self[leftChild] = self[leftChild],self[i]
                    self.shiftdown(leftChild,end)
                else:
                    self[i],self[rightChild] = self[rightChild],self[i]
                    self.shiftdown(rightChild,end)
            return

    def  quickSort(self,first,last):
        
        if (first >= last):
            return
        
        pivot , left , right = self[first] , first + 1 , last
        while (left < right):
            while (left < right and self[left] < pivot):
                left += 1
            while (right > left and self[right] >= pivot):
                right -= 1
            self[left],self[right] = self[right],self[left]
            
        '''if (self[left] < pivot):
            center = left
        else:
            center = left - 1'''
        center = left if self[left] < pivot else left - 1
        self[first],self[center] = self[center],self[first]
        self.quickSort(first , center - 1)
        self.quickSort(center + 1,last)
        return
    
    def  quickSort_randint(self,first,last):
        
        if (first >= last):
            return
        pivot , left , right = self[random.randint(first+1,last-1)] , first + 1 , last
        while (left < right):
            while (left < right and self[left] < pivot):
                left += 1
            while (right > left and self[right] >= pivot):
                right -= 1
            self[left],self[right] = self[right],self[left]
        center = left if self[left] < pivot else left - 1
        self[first],self[center] = self[center],self[first]
        self.quickSort(first , center - 1)
        self.quickSort(center + 1,last)
        return
        
    def heapSortNonRecur(x):
        '''HeapSort Non-Recursive Version'''
    
        length = len(x)
        if length <= 1:
            return
    
        '''Heapify'''
        for i in range(length):
            j = i
            while j > 0 and x[j] < x[(j-1)//2]:
                x[j], x[(j-1)//2] = x[(j-1)//2], x[j]
                j = (j-1)//2
            
        for i in range(length-1, 0, -1):
            x[i], x[0] = x[0], x[i]
        
            '''Shift Down'''
            j = 0
            last = i-1
            while j * 2 + 1 <= last: # as long as j has child
                lc = j * 2 + 1
                rc = j * 2 + 2
                if lc == last: # There is only one child (left child)
                    if x[j] > x[lc]:
                        x[j], x[lc] = x[lc], x[j]
                        j = lc
                    else:
                        break
                else: # Consider both children
                    if x[j] > min(x[lc], x[rc]):
                        if x[lc] < x[rc]:
                            x[j], x[lc] = x[lc], x[j]
                            j = lc
                        else:
                            x[j], x[rc] = x[rc], x[j]
                            j = rc
                    else:
                        break
        return

# The report of comparing different sort algroithms


## Contents
* .sort() in python
* Selected sort
* Bubble sort
* Merge sort
* Heap sort
* Quick sort
* Summary
* Reference


### .sort() in python

In [54]:

random.seed(2018)
testList = [random.randint(0, 100000) for i in range(100)]
t = time.time()
testList.sort()
print("Cost time        (100): ", time.time()-t)
testList = [random.randint(0, 100000) for i in range(1000)]
t = time.time()
testList.sort()
print("Cost time         (1K): ", time.time()-t)
testList = [random.randint(0, 100000) for i in range(10000)]
t = time.time()
testList.sort()
print("Cost time        (10K): ", time.time()-t)
testList = [random.randint(0, 100000) for i in range(100000)]
t = time.time()
testList.sort()
print("Cost time       (100K): ", time.time()-t)
sortedList = [random.randint(0, 100000) for i in range(1000000)]
t = time.time()
sortedList.sort()
print("Cost time         (1M): ", time.time()-t)
testList = [random.randint(0, 100000) for i in range(10000000)]
t = time.time()
testList.sort()
print("Cost time        (10M): ", time.time()-t)

t = time.time()
sortedList.sort()
print("Cost time(sorted 1M)  : ", time.time()-t)

Cost time        (100):  0.0
Cost time         (1K):  0.0
Cost time        (10K):  0.0029561519622802734
Cost time       (100K):  0.03191494941711426
Cost time         (1M):  0.49255990982055664
Cost time        (10M):  6.293714284896851
Cost time(sorted 1M)  :  0.049854278564453125


我使用了seed(2018)以確保每次被排序的數列都一樣，而已經排序好的數列我選擇1M

### Selected sort
Selection sort is another simple sorting method that works better than bubble sort It works by finding the smallest or highest element
(most probably desired element) from the unsorted list and swaps with the first element
in the sorted list and then finds the next smallest element from the unsorted
list and swap with the second element in the sorted list.
#### Explanation 
An arry of elements A [154, 126, 193, 117, 177, 131, 144, 155, 120]
and sorting the array in ascending order using selection sort.
![Alt text](https://upload.cc/i1/2018/10/02/SDeubr.png)
In the above figure the algorithm finds the largest element in the list and swaps
it with the appropriate position.The algorithm continues this operation until the list being sorted.

In [4]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.selectSort()
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.selectSort()
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.selectSort()
print("Cost time        (10k) : ", time.time()-t)
'''testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.selectSort()
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.selectSort()
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.selectSort()
print("Cost time        (10M) : ", time.time()-t)
t = time.time()'''
'''sortedList.selectSort()
print("Cost time(sorted 1M)   : ", time.time()-t)'''

Cost time        (100) :  0.001994609832763672
Cost time        (1k)  :  0.14223885536193848
Cost time        (10k) :  14.580124139785767


'sortedList.selectSort()\nprint("Cost time(sorted 1M)   : ", time.time()-t)'

SelectSort當內容超過10k時，就會跑超過5分鐘，畢竟其為O(n^2)的演算法，當內容增加時其複雜度呈現指數成長，因此為一沒有效率的演算法

### Bubble sort
Bubble sort is a simple and the slowest sorting algorithm which works by comparing each element in the list with its neighboring elements and swapping them if they are in undesirable order.The algorithm continues this operation until it makes a pass right through the list without swapping any elements, which shows that the
list is sorted. This process makes the algorithm works slower when the size of the input n increased. 

#### Explanation

Consider an arry of elements A[5,7,3,1,0,7,2]  
First pass:  
[**5**,**7**,3,1,0,7,2] -> [**5**,**7**,3,1,0,7,2] Nothing changed in this step  
Second pass:  
[5,**7**,**3**,1,0,7,2] -> [5,**3**,**7**,1,0,7,2] Algorithm swap 3 and 7 seeing 7 > 3   
[5,3,**7**,**1**,0,7,2] -> [5,3,**1**,**7**,0,7,2] Algorithm swap 1 and 7 seeing 7 > 1   
[5,3,1,**7**,**0**,7,2] -> [5,3,1,**0**,**7**,7,2] Algorithm swap 0 and 7 seeing 7 > 0  
[5,3,1,0,**7**,**7**,2] -> [5,3,1,0,**7**,**7**,2] Algorithm swap 7 and 7 seeing 7 = 7  
[5,3,1,0,7,**7**,**2**] -> [5,3,1,0,7,**2**,**7**] Algorithm swap 2 and 7 seeing 7 > 2  
The process contiune until the list been sorted.

In [5]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.bubbleSort()
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.bubbleSort()
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.bubbleSort()
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
'''t = time.time()
testList.bubbleSort()
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.bubbleSort()
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.bubbleSort()
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.bubbleSort()
print("Cost time(sorted 1M)   : ", time.time()-t)'''

Cost time        (100) :  0.0009963512420654297
Cost time        (1k)  :  0.10835027694702148
Cost time        (10k) :  12.669689416885376


't = time.time()\ntestList.bubbleSort()\nprint("Cost time        (100k): ", time.time()-t)\nsortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])\nt = time.time()\nsortedList.bubbleSort()\nprint("Cost time        (1M)  : ", time.time()-t)\ntestList = NTHUList([random.randint(0, 100000) for i in range(10000000)])\nt = time.time()\ntestList.bubbleSort()\nprint("Cost time        (10M) : ", time.time()-t)\nt = time.time()\nsortedList.bubbleSort()\nprint("Cost time(sorted 1M)   : ", time.time()-t)'

### Merge sort
Merge sort works by splitting the unsorted array into n sub array recursively until each sub
array has 1 element.Consequently it merges each sub array to generate a final sorted array.
#### Explation
Consider a array of elements A [54, 26, 93, 17, 77, 31, 44, 55] and sorting the
array into ascending order using merge sort.
![Alt text](https://upload.cc/i1/2018/10/02/CigPuW.png)
Merge sort divides the unsorted array into two equal sub arrays and repeats this operation until each sub array has 1 elements. Now each sub array considered
to be sorted and the merging process proceeds. The merge operation scans the two
sub array and finds the smallest element and thus removes the smallest element and
placed in the final sorted sub array. This operation is repeated until there is only
one sub array left. This sub array considered to be sorted.

In [42]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.mergeSort()
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.mergeSort()
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.mergeSort()
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.mergeSort()
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.mergeSort()
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.mergeSort()
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.mergeSort()
print("Cost time(sorted 1M)   : ", time.time()-t)

Cost time        (100) :  0.0
Cost time        (1k)  :  0.0039882659912109375
Cost time        (10k) :  0.049866676330566406
Cost time        (100k):  0.629314661026001
Cost time        (1M)  :  7.711877107620239
Cost time        (10M) :  89.72606205940247
Cost time(sorted 1M)   :  7.284044504165649


Merge sort 使用 divide and conquer 的方法來實作，缺點是在使用上會需要額外的空間來處理

### Heap sort
In general, heap is a specialized tree based data structure that satisfies the heap
property, mostly we use binary tree. The tree structure is well balanced, space
efficient and fast. Heap sort works by building a heap from the input array and
then removing the maximum element from the heap and placing it at the end of the
final sorted array 
#### Explanation
Let us consider an array of elements A [16, 14, 10, 8, 7, 9, 3, 2 4, 1] and sorting the
array into ascending order using heap sort.
![Alt text](https://upload.cc/i1/2018/10/02/dgPzGb.png)
 The algorithm removes the maximum element from the
tree and placed in the rightmost position in the final sorted array. We can see each
time when the maximum element is removed the tree restores its heap property.The heap property states that every node should has the higher value than its child
node.The algorithm repeats this operation until the array being
sorted. 

In [43]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.heapSort()
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.heapSort()
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.heapSort()
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.heapSort()
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.heapSort()
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.heapSort()
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.heapSort()
print("Cost time(sorted 1M)   : ", time.time()-t)

Cost time        (100) :  0.000997781753540039
Cost time        (1k)  :  0.012965917587280273
Cost time        (10k) :  0.14165186882019043
Cost time        (100k):  1.8234925270080566
Cost time        (1M)  :  22.23289179801941
Cost time        (10M) :  281.5308835506439
Cost time(sorted 1M)   :  28.130272388458252


heap sort 以linked list 的方式儲存資料使其不需要額外記憶體做處裡，但為人詬病的是其速度相較於同複雜度的演算法，總是較慢，例如heap sort 和 merge sort 的複雜度雖然都是 nlogn 但是在此實驗中heap sort的時間明顯高於 merge sort ，我推測是因為heap sort 呼叫函式拖慢了速度

### HeapSortNonRecur

In [5]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.heapSortNonRecur()
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.heapSortNonRecur()
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.heapSortNonRecur()
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.heapSortNonRecur()
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.heapSortNonRecur()
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.heapSortNonRecur()
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.heapSortNonRecur()
print("Cost time(sorted 1M)   : ", time.time()-t)

Cost time        (100) :  0.0010046958923339844
Cost time        (1k)  :  0.009971141815185547
Cost time        (10k) :  0.12466883659362793
Cost time        (100k):  1.874927282333374
Cost time        (1M)  :  22.595292568206787
Cost time        (10M) :  297.1841766834259
Cost time(sorted 1M)   :  29.339892864227295


### Quick sort
Quick sort  is the fastest general purpose internal sorting algorithm on the average
among other sophisticated algorithms. Unlike merge sort it does not require any
additional memory space for sorting an array.It works by
selecting elements from unsorted array named as a pivot and split the array into two
parts called sub arrays and reconstruct the former part with the elements smaller
than the pivot and the latter with elements larger than the pivot. This operation
is called as partitioning. The algorithm repeats this operation recursively for both
the sub arrays.
#### Explanation
Let us consider an array of elements A [3, 7, 8, 5, 2, 1, 9, 5, 4] and sorting the array
into ascending order using quick sort.
![Alt text](https://upload.cc/i1/2018/10/02/So3JaB.png)
In the above figure, the algorithm first selects the random element as a pivot
and hides the pivot in the next meanwhile it reconstructs the array with elements
smaller than pivot in the left sub array, likewise elements larger than pivot in the
right sub array. The algorithm repeats this operation until the array being sorted.

In [51]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.quickSort(0,len(testList)-1)
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.quickSort(0,len(testList)-1)
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.quickSort(0,len(testList)-1)
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.quickSort(0,len(testList)-1)
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.quickSort(0,len(testList)-1)
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.quickSort(0,len(testList)-1)
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.quickSort(0,len(sortedList)-1)
print("Cost time(sorted 1M)   : ", time.time()-t)

Cost time        (100) :  0.0
Cost time        (1k)  :  0.0029611587524414062
Cost time        (10k) :  0.03989291191101074
Cost time        (100k):  0.4735379219055176
Cost time        (1M)  :  0.46468019485473633
Cost time        (10M) :  136.33708143234253
Cost time(sorted 1M)   :  7.2985029220581055


為了比較，我另外使用pivot 為隨機數版本

In [52]:
random.seed(2018)
testList = NTHUList([random.randint(0, 100000) for i in range(100)])
t = time.time()
testList.quickSort_randint(0,len(testList)-1)
print("Cost time        (100) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(1000)])
t = time.time()
testList.quickSort_randint(0,len(testList)-1)
print("Cost time        (1k)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000)])
t = time.time()
testList.quickSort_randint(0,len(testList)-1)
print("Cost time        (10k) : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(100000)])
t = time.time()
testList.quickSort_randint(0,len(testList)-1)
print("Cost time        (100k): ", time.time()-t)
sortedList = NTHUList([random.randint(0, 100000) for i in range(1000000)])
t = time.time()
sortedList.quickSort_randint(0,len(testList)-1)
print("Cost time        (1M)  : ", time.time()-t)
testList = NTHUList([random.randint(0, 100000) for i in range(10000000)])
t = time.time()
testList.quickSort_randint(0,len(testList)-1)
print("Cost time        (10M) : ", time.time()-t)
t = time.time()
sortedList.quickSort_randint(0,len(sortedList)-1)
print("Cost time(sorted 1M)   : ", time.time()-t)

Cost time        (100) :  0.0
Cost time        (1k)  :  0.0029871463775634766
Cost time        (10k) :  0.040884971618652344
Cost time        (100k):  0.5165872573852539
Cost time        (1M)  :  0.45378637313842773
Cost time        (10M) :  162.2275538444519
Cost time(sorted 1M)   :  7.778953790664673


## Summary
可以看出bubble 和 selection 在data超過10K時基本跑不了，而表現最好的是python內建的sort()，我的推論是可能它有對python這個語言進行特別加強。heapsort的表現不如merge好，不過其不占記憶體的特性仍是不可忽視的。Quick sort 無疑問的為最快速的演算法，雖然他在10M時的表現不如merge，但我猜測這只是偶發的現象。我原本以為使用nonrecur version的 heapsort 會表現更加，但實驗似乎顯示recur version的更好。
### 100
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.0 <br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : 0.000997304916381836 <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : 0.002026081085205078 <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 0.0 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;: 0.000997781753540039 <br>
heapSortNonRecur : 0.0010046958923339844 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 0.0 <br>
quick sort_randint&nbsp;                                                            : 0.0 <br>

### 1K
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.0 <br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : 0.1475822925567627 <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : 0.13184523582458496 <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 0.0039882659912109375 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;: 0.012965917587280273 <br>
heapSortNonRecur : 0.009971141815185547 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 0.0029611587524414062 <br>
quick sort_randint&nbsp;                                                            : 0.0029871463775634766 <br>
### 10K
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.0020017623901367188 <br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : 14.949881315231323 <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : 14.510718822479248 <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           :  0.049866676330566406 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;:  0.14165186882019043 <br>
heapSortNonRecur : 0.12466883659362793 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 0.03989291191101074 <br>
quick sort_randint&nbsp;    :   0.040884971618652344
### 100K
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.028922557830810547 <br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : RT <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : RT <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 0.629314661026001 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;:  1.8234925270080566 <br>
heapSortNonRecur : 1.874927282333374 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 0.4735379219055176 <br>
quick sort_randint&nbsp;    :   0.5165872573852539
### 1M
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.4721090793609619 <br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : RT <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : RT <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 7.711877107620239 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;:  22.23289179801941 <br>
heapSortNonRecur : 22.595292568206787 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 0.46468019485473633 <br>
quick sort_randint&nbsp;    :    0.45378637313842773
### 10M
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 5.860732555389404<br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : RT <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : RT <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 89.72606205940247 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;:  281.5308835506439 <br>
heapSortNonRecur: 297.1841766834259 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 136.33708143234253 <br>
quick sort_randint&nbsp;    :    162.2275538444519
### sorted
.sort() in python&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;             : 0.04919028282165527<br>
selected sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;           : RT <br>
bubble sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;       : RT <br>
merge sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;  &nbsp;&nbsp;           : 7.284044504165649 <br>
heap sort &nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp; &nbsp;&nbsp;&nbsp;:  28.130272388458252 <br>
heapSortNonRecur : 29.339892864227295 <br>
quick sort&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;&nbsp;: 7.2985029220581055 <br>
quick sort_randint&nbsp;    :    7.778953790664673

Note : RT 表示時間過久

## Reference
Ashok Kumar Karunanithi(2014)。A Survey, Discussion and Comparison of Sorting Algorithms。
